### PR TITLE
Test Azure CSI storage

### PR DESCRIPTION
**Pull request by sebsoto**
Adds azure-file CSI storage to the e2e tests

### CODE REVIEW

File: test/e2e/providers/azure/azure.go [hunk 1]
```diff
@@ -119,11 +124,218 @@ func (p *Provider) GetType() config.PlatformType {
...
+	}
+	storageClassName := "azurefile-csi"
+	pvcSpec := core.PersistentVolumeClaim{
+		ObjectMeta: meta.ObjectMeta{
+			GenerateName: "e2e" + "-",
+		},
+		Spec: core.PersistentVolumeClaimSpec{
+			AccessModes: []core.PersistentVolumeAccessMode{core.ReadWriteMany},
+			Resources: core.ResourceRequirements{
+				Requests: core.ResourceList{core.ResourceStorage: resource.MustParse("2Gi")},
```
Comment (saifshaikh48) [inline test/e2e/providers/azure/azure.go:142]: nit: do we want to request less storage for the e2e suite?

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

File: test/e2e/providers/azure/azure.go (hunk 1 above)
Comment (alinaryan) [reply test/e2e/providers/azure/azure.go:142]: I thought 2Gi was already pretty small. Also not sure how much customers will have.

**Review by alinaryan — COMMENTED** [review]
Thanks for working on this, are we going to worry about/test Azure-Disk too?

**Comment by sebsoto** [comment]
> are we going to worry about/test Azure-Disk too?

I don't think thats necessary in CI for now, maybe eventually

File: test/e2e/providers/azure/azure.go (hunk 1 above)
Comment (aravindhp) [reply test/e2e/providers/azure/azure.go:142]: 2Gi is small in today's world 😄

**Review by aravindhp — APPROVED** [review]

File: test/e2e/providers/azure/azure.go [hunk 2]
```diff
@@ -119,11 +124,218 @@ func (p *Provider) GetType() config.PlatformType {
...
+						{
+							Key:    "os",
+							Value:  "Windows",
+							Effect: core.TaintEffectNoSchedule,
+						},
+					},
+					Containers: []core.Container{
+						{
+							Name:  "node-driver-registrar",
+							Image: "mcr.microsoft.com/oss/kubernetes-csi/csi-node-driver-registrar:v2.8.0",
```
Comment (jrvaldes) [inline test/e2e/providers/azure/azure.go:181]: Why not using the [latest](https://github.com/kubernetes-csi/node-driver-registrar/releases/tag/v2.9.0) (v2.9.0) as of now?

File: test/e2e/providers/azure/azure.go [hunk 3]
```diff
@@ -119,11 +124,218 @@ func (p *Provider) GetType() config.PlatformType {
...
+								},
+								{
+									Name:      "registration-dir",
+									MountPath: "/registration",
+								},
+							},
+						},
+						{
+							Name:  "azurefile",
+							Image: "mcr.microsoft.com/k8s/csi/azurefile-csi:latest",
```
Comment (jrvaldes) [inline test/e2e/providers/azure/azure.go:218]: Should we worry about matching compatible image versions of the `azurefile-csi` and `csi-node-driver-registrar`?

File: test/e2e/providers/azure/azure.go [hunk 4]
```diff
@@ -119,11 +124,218 @@ func (p *Provider) GetType() config.PlatformType {
...
+			},
+		},
+	}
+	// See if the DaemonSet already exists in the state we expect it to be in.
+	existingDS, err := client.AppsV1().DaemonSets(csiNamespace).Get(context.TODO(), dsName, meta.GetOptions{})
+	if err != nil {
+		if !k8sapierrors.IsNotFound(err) {
+			return fmt.Errorf("error getting existing Windows CSI DaemonSet: %w", err)
+		}
+	} else if err == nil {
```
Comment (jrvaldes) [inline test/e2e/providers/azure/azure.go:322]: No need for `if err == nil`

File: test/e2e/providers/azure/azure.go [hunk 5]
```diff
@@ -119,11 +124,218 @@ func (p *Provider) GetType() config.PlatformType {
...
+		if reflect.DeepEqual(existingDS.Spec, ds.Spec) {
+			// DaemonSet is already as expected, nothing to do here.
+			return nil
+		}
+		// Delete the DaemonSet as it has the wrong spec.
+		err = client.AppsV1().DaemonSets(csiNamespace).Delete(context.TODO(), dsName, meta.DeleteOptions{})
+		if err != nil {
+			return fmt.Errorf("error deleting existing Windows CSI DaemonSet: %w", err)
+
+		}
```
Comment (jrvaldes) [inline test/e2e/providers/azure/azure.go:332]: Should we wait for the existing DaemonSet to be deleted before trying to create one?

File: test/e2e/providers/azure/azure.go [hunk 6]
```diff
@@ -119,11 +124,218 @@ func (p *Provider) GetType() config.PlatformType {
...
+	}
+	_, err = client.AppsV1().DaemonSets(csiNamespace).Create(context.TODO(), &ds, meta.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating Windows CSI DaemonSet: %w", err)
+	}
+	return nil
```
Comment (jrvaldes) [inline test/e2e/providers/azure/azure.go:338]: Do you think we should check the containers are running before returning?

**Review by jrvaldes — COMMENTED** [review]
@sebsoto  thanks for working on this. I left minor comments. LGTM otherwise.

**Comment by sebsoto** [comment]
/test azure-e2e-operator
Cluster come up failure

File: test/e2e/providers/azure/azure.go (hunk 2 above)
Comment (sebsoto) [reply test/e2e/providers/azure/azure.go:181]: This is following the documentation provided https://github.com/kubernetes-sigs/azurefile-csi-driver/tree/master/deploy

File: test/e2e/providers/azure/azure.go (hunk 4 above)
Comment (sebsoto) [reply test/e2e/providers/azure/azure.go:322]: It is needed in order to exclude the case where `k8sapierrors.IsNotFound(err)` evaluates to true

File: test/e2e/providers/azure/azure.go (hunk 3 above)
Comment (sebsoto) [reply test/e2e/providers/azure/azure.go:218]: This is following the documentation provided https://github.com/kubernetes-sigs/azurefile-csi-driver/tree/master/deploy

File: test/e2e/providers/azure/azure.go (hunk 4 above)
Comment (jrvaldes) [reply test/e2e/providers/azure/azure.go:322]: That case is excluded by the `else` statement. `err == nil` is always true in the block following the `else` in
```
if err != nil {
	
} else {

}
```

Comment (sebsoto) [reply test/e2e/providers/azure/azure.go:322]: yep you are right, fixed

File: test/e2e/providers/azure/azure.go (hunk 5 above)
Comment (sebsoto) [reply test/e2e/providers/azure/azure.go:332]: No need, the pods will do the correct thing

File: test/e2e/providers/azure/azure.go (hunk 6 above)
Comment (sebsoto) [reply test/e2e/providers/azure/azure.go:338]: I don't think that's necessary, if the containers don't go to running the workload won't be able to deploy so the failure will be shown there.

**Comment by sebsoto** [comment]
Overriding due to vsphere test flakes, and the fact that this PR will only affect azure clusters

**Pull request merged**